File: momentra/momentra/Shell/GroupActive/GroupActionRegistry.swift
import Foundation
import SwiftUI

/// V019 Group Quick Add + Figma 575:14655 tile catalog. Mirrors Android `GroupActionRegistry`.
enum GroupActionCode: String, CaseIterable, Equatable {
    case expenseCreate = "EXPENSE_CREATE"
    case contributionRecord = "CONTRIBUTION_RECORD"
    case settlementRecord = "SETTLEMENT_RECORD"
    case participantManage = "PARTICIPANT_MANAGE"
    case planningItemCreate = "PLANNING_ITEM_CREATE"
    case bookingCreate = "BOOKING_CREATE"
    case pollCreate = "POLL_CREATE"
    case updateCreate = "UPDATE_CREATE"
    case memoryCreate = "MEMORY_CREATE"
    case purchaseItemCreate = "PURCHASE_ITEM_CREATE"
    case residentManage = "RESIDENT_MANAGE"
}

enum GroupActionDestination: Equatable {
    case expense
    case contribution
    case settlement
    case participants
    case budget
    case planning
    case booking
    case poll
    case memory
    case update
    case invite
    case purchaseItem
    case resident
}

struct GroupActionTile: Identifiable {
    var id: String { "\(code?.rawValue ?? tileId)-\(label)" }
    let code: GroupActionCode?
    let tileId: String
    let label: String
    let subtitle: String
    let icon: String
    let colors: [Color]
    let destination: GroupActionDestination
    let enabledWhenMomentActive: Bool
    let apiGap: Bool
}

enum GroupActionRegistry {
    static func destination(for capabilityCode: String) -> GroupActionDestination? {
        switch capabilityCode.uppercased() {
        case GroupActionCode.expenseCreate.rawValue: return .expense
        case GroupActionCode.contributionRecord.rawValue: return .contribution
        case GroupActionCode.settlementRecord.rawValue: return .settlement
        case GroupActionCode.participantManage.rawValue: return .participants
        case GroupActionCode.planningItemCreate.rawValue: return .planning
        case GroupActionCode.bookingCreate.rawValue: return .booking
        case GroupActionCode.pollCreate.rawValue: return .poll
        case GroupActionCode.updateCreate.rawValue: return .update
        case GroupActionCode.memoryCreate.rawValue: return .memory
        case GroupActionCode.purchaseItemCreate.rawValue: return .purchaseItem
        case GroupActionCode.residentManage.rawValue: return .resident
        default: return nil
        }
    }

    static func figmaHubTiles(
        hasActiveMoment: Bool,
        capabilityCodes: [String]? = nil
    ) -> [GroupActionTile] {
        catalogTiles().map { tile in
            let enabled = hubTileEnabled(hasActiveMoment: hasActiveMoment, capabilityCodes: capabilityCodes, tile: tile)
            return GroupActionTile(
                code: tile.code,
                tileId: tile.tileId,
                label: tile.label,
                subtitle: tile.subtitle,
                icon: tile.icon,
                colors: tile.colors,
                destination: tile.destination,
                enabledWhenMomentActive: enabled,
                apiGap: tile.apiGap
            )
        }
    }

    /// Legacy API for tests.
    static func tiles(hasActiveMoment: Bool, capabilityCodes: [String]? = nil) -> [GroupActionTile] {
        figmaHubTiles(hasActiveMoment: hasActiveMoment, capabilityCodes: capabilityCodes)
    }

    private static func hubTileEnabled(
        hasActiveMoment: Bool,
        capabilityCodes: [String]?,
        tile: GroupActionTile
    ) -> Bool {
        if !hasActiveMoment { return false }
        if tile.apiGap { return false }
        // Empty capabilities must NOT enable everything — fail closed until bootstrap fills V019.
        guard let codes = capabilityCodes, !codes.isEmpty else {
            return tile.code == nil
        }
        guard let code = tile.code?.rawValue else { return true }
        return codes.map { $0.uppercased() }.contains(code.uppercased())
    }

    private static func catalogTiles() -> [GroupActionTile] {
        [
            tile("expense", .expenseCreate, "Expense", "Split a cost", "creditcard.fill", "#33C759", "#0F766E", .expense),
            tile("planning", .planningItemCreate, "Planning", "Itinerary & tasks", "calendar", "#14B8A6", "#0F766E", .planning),
            tile("budget", nil, "Budget", "Edit planned total", "chart.bar.fill", "#FFB598", "#E8621A", .budget),
            tile("booking", .bookingCreate, "Booking", "Reservations", "ticket.fill", "#FF7A3D", "#E85940", .booking),
            tile("poll", .pollCreate, "Poll", "Group decisions", "checklist", "#A855F7", "#7C3AED", .poll),
            tile("memory", .memoryCreate, "Memory", "Capture a moment", "camera.fill", "#FF8E63", "#E8744F", .memory),
            tile("update", .updateCreate, "Update", "Share status", "megaphone.fill", "#3B82F6", "#1D4ED8", .update),
            tile("contribution", .contributionRecord, "Contribution", "Record money in", "hand.raised.fill", "#10B981", "#047857", .contribution),
            tile("invite", .participantManage, "Invite", "Add people", "person.badge.plus", "#FFB598", "#E8621A", .invite),
            tile("settle", .settlementRecord, "Settle", "Pay down balances", "scalemass", "#059669", "#10B981", .settlement),
            tile("purchase", .purchaseItemCreate, "Purchase item", "Track a buy", "cart.fill", "#F59E0B", "#D97706", .purchaseItem),
            tile("resident", .residentManage, "Resident", "Add a housemate", "house.fill", "#6366F1", "#4338CA", .resident),
        ]
    }

    private static func tile(
        _ tileId: String,
        _ code: GroupActionCode?,
        _ label: String,
        _ subtitle: String,
        _ icon: String,
        _ start: String,
        _ end: String,
        _ destination: GroupActionDestination,
        apiGap: Bool = false
    ) -> GroupActionTile {
        GroupActionTile(
            code: code,
            tileId: tileId,
            label: label,
            subtitle: subtitle,
            icon: icon,
            colors: [Color(hex: start), Color(hex: end)],
            destination: destination,
            enabledWhenMomentActive: !apiGap,
            apiGap: apiGap
        )
    }

    static func equalSplitInputs(participantIds: [String]) -> [APIClient.GroupSplitInput] {
        participantIds.map { APIClient.GroupSplitInput(participantId: $0) }
    }

    static func defaultCodes() -> [GroupActionCode] {
        GroupActionCode.allCases
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/group/GroupActionRegistry.kt
package com.example.momentra.ui.shell.group

import com.example.momentra.R

/**
 * Mapper from V019 Group capability / action codes to Quick Add destinations.
 * SETTLEMENT_RECORD is LIVE (V047) — enable Settle when capability present.
 */
object GroupActionRegistry {

    const val EXPENSE_CREATE = "EXPENSE_CREATE"
    const val CONTRIBUTION_RECORD = "CONTRIBUTION_RECORD"
    const val SETTLEMENT_RECORD = "SETTLEMENT_RECORD"
    const val PARTICIPANT_MANAGE = "PARTICIPANT_MANAGE"
    const val PLANNING_ITEM_CREATE = "PLANNING_ITEM_CREATE"
    const val BOOKING_CREATE = "BOOKING_CREATE"
    const val POLL_CREATE = "POLL_CREATE"
    const val UPDATE_CREATE = "UPDATE_CREATE"
    const val MEMORY_CREATE = "MEMORY_CREATE"
    const val PURCHASE_ITEM_CREATE = "PURCHASE_ITEM_CREATE"
    const val RESIDENT_MANAGE = "RESIDENT_MANAGE"

    enum class Destination {
        EXPENSE,
        CONTRIBUTION,
        SETTLEMENT,
        PARTICIPANTS,
        BUDGET,
        PLANNING,
        BOOKING,
        POLL,
        MEMORY,
        UPDATE,
        INVITE,
        PURCHASE_ITEM,
        RESIDENT,
    }

    data class HubTileSpec(
        val id: String,
        val label: String,
        val subtitle: String,
        val iconRes: Int,
        val gradientStart: androidx.compose.ui.graphics.Color,
        val gradientEnd: androidx.compose.ui.graphics.Color,
        val destination: Destination,
        val capabilityCode: String? = null,
        val apiGap: Boolean = false,
    )

    /** Figma 575:14655 — Trip Action Center 3×3 grid. */
    val tripHubTileIds: Set<String> = setOf(
        "expense", "planning", "budget", "booking", "poll", "memory", "update", "contribution", "invite",
    )

    val figmaHubTiles: List<HubTileSpec> = listOf(
        HubTileSpec("expense", "Expense", "Split a cost", R.drawable.ic_group_qa_wallet, androidx.compose.ui.graphics.Color(0xFF33C759), androidx.compose.ui.graphics.Color(0xFF0F766E), Destination.EXPENSE, EXPENSE_CREATE),
        HubTileSpec("planning", "Planning", "Itinerary & tasks", R.drawable.ic_group_qa_calendar, androidx.compose.ui.graphics.Color(0xFF14B8A6), androidx.compose.ui.graphics.Color(0xFF0F766E), Destination.PLANNING, PLANNING_ITEM_CREATE),
        HubTileSpec("budget", "Budget", "Edit planned total", R.drawable.ic_group_qa_chartbar, androidx.compose.ui.graphics.Color(0xFFFFB598), androidx.compose.ui.graphics.Color(0xFFE8621A), Destination.BUDGET),
        HubTileSpec("booking", "Booking", "Reservations", R.drawable.ic_group_qa_ticket, androidx.compose.ui.graphics.Color(0xFFFF7A3D), androidx.compose.ui.graphics.Color(0xFFE85940), Destination.BOOKING, BOOKING_CREATE),
        HubTileSpec("poll", "Poll", "Group decisions", R.drawable.ic_group_qa_vote, androidx.compose.ui.graphics.Color(0xFFA855F7), androidx.compose.ui.graphics.Color(0xFF7C3AED), Destination.POLL, POLL_CREATE),
        HubTileSpec("memory", "Memory", "Capture a moment", R.drawable.ic_group_qa_camera, androidx.compose.ui.graphics.Color(0xFFFF8E63), androidx.compose.ui.graphics.Color(0xFFE8744F), Destination.MEMORY, MEMORY_CREATE),
        HubTileSpec("update", "Update", "Share status", R.drawable.ic_group_qa_megaphone, androidx.compose.ui.graphics.Color(0xFF3B82F6), androidx.compose.ui.graphics.Color(0xFF1D4ED8), Destination.UPDATE, UPDATE_CREATE),
        HubTileSpec("contribution", "Contribution", "Record money in", R.drawable.ic_group_qa_handshake, androidx.compose.ui.graphics.Color(0xFF10B981), androidx.compose.ui.graphics.Color(0xFF047857), Destination.CONTRIBUTION, CONTRIBUTION_RECORD),
        HubTileSpec("invite", "Invite", "Add people", R.drawable.ic_group_qa_userplus, androidx.compose.ui.graphics.Color(0xFFFFB598), androidx.compose.ui.graphics.Color(0xFFE8621A), Destination.INVITE, PARTICIPANT_MANAGE),
        HubTileSpec("settle", "Settle", "Pay down balances", R.drawable.ic_group_qa_wallet, androidx.compose.ui.graphics.Color(0xFF059669), androidx.compose.ui.graphics.Color(0xFF10B981), Destination.SETTLEMENT, SETTLEMENT_RECORD),
        HubTileSpec("purchase", "Purchase item", "Track a buy", R.drawable.ic_group_qa_chartbar, androidx.compose.ui.graphics.Color(0xFFF59E0B), androidx.compose.ui.graphics.Color(0xFFD97706), Destination.PURCHASE_ITEM, PURCHASE_ITEM_CREATE),
        HubTileSpec("resident", "Resident", "Add a housemate", R.drawable.ic_group_qa_userplus, androidx.compose.ui.graphics.Color(0xFF6366F1), androidx.compose.ui.graphics.Color(0xFF4338CA), Destination.RESIDENT, RESIDENT_MANAGE),
    )

    val figmaTripHubTiles: List<HubTileSpec>
        get() = figmaHubTiles.filter { it.id in tripHubTileIds }

    fun destinationFor(capabilityCode: String): Destination? = when (capabilityCode.uppercase()) {
        EXPENSE_CREATE -> Destination.EXPENSE
        CONTRIBUTION_RECORD -> Destination.CONTRIBUTION
        SETTLEMENT_RECORD -> Destination.SETTLEMENT
        PARTICIPANT_MANAGE -> Destination.PARTICIPANTS
        PLANNING_ITEM_CREATE -> Destination.PLANNING
        BOOKING_CREATE -> Destination.BOOKING
        POLL_CREATE -> Destination.POLL
        UPDATE_CREATE -> Destination.UPDATE
        MEMORY_CREATE -> Destination.MEMORY
        PURCHASE_ITEM_CREATE -> Destination.PURCHASE_ITEM
        RESIDENT_MANAGE -> Destination.RESIDENT
        else -> null
    }

    fun isDestinationEnabled(capabilities: List<String>, destination: Destination): Boolean {
        val tile = figmaHubTiles.find { it.destination == destination } ?: return false
        if (tile.apiGap) return false
        // Empty capabilities must NOT enable everything — fail closed until bootstrap fills V019 codes.
        if (capabilities.isEmpty()) return tile.capabilityCode == null
        return tile.capabilityCode?.let { code ->
            capabilities.any { it.equals(code, ignoreCase = true) }
        } ?: true
    }

    fun hubTileEnabled(hasActiveMoment: Boolean, capabilities: List<String>, tile: HubTileSpec): Boolean {
        if (!hasActiveMoment) return false
        if (tile.apiGap) return false
        if (capabilities.isEmpty()) return tile.capabilityCode == null
        return tile.capabilityCode?.let { code ->
            capabilities.any { it.equals(code, ignoreCase = true) }
        } ?: true
    }

    fun enabledDestinations(capabilities: List<String>): Set<Destination> {
        return figmaHubTiles.filter { tile ->
            hubTileEnabled(hasActiveMoment = true, capabilities = capabilities, tile = tile)
        }.map { it.destination }.toSet()
    }
}
